### ISSUE TITLE
[RFC] Move adapters into the 'core' repository

### ISSUE THREAD

**Issue by sstok** [gushphp/gush#489]
All adapters are at this moment housed in there own repositories, which is good from a package design point-of-view. But the reality is that they are not really optional, the gush core package already requires them and we don't distribute Gush with Composer anymore (only a phar).
*And this circular dependency between the packages is also not good package design.*

Whenever we update an adapter we already need to recreate a new phar, and make sure the adapters are up-to-date (which is something we forget to often).

If we move the adapters to the core repository:

We can easily change the API (without having to worry about the tagging of a version or having to delay a release as not all adapters are compatible yet, which could also be solved with a better release management but still is no easy task).

Cons:

* *If* we find a good way to enable plug-ins we are stuck with all adapters, which means we may need to move them back to there own repositories again.
* Possible loss of commit authorship (is this really a problem?)
* Small fixes in an adapter will require a new tag; but at this moment this already requires a new release.

The problem of Composer version requirements when working on a new feature is already resolved, that is no problem anymore.

**In conclusion**

We can either move all adapters into the core repository and have less to worry about, or be very strict about the versioning and make sure the required adapter versions are up-to-date and compatible whenever we want to release.

Also, you need at least one Adapter enabled as there are no commands anymore that don't require the adapter (except for `core:*`, and `meta:header` (which could be extracted to its own mini tool)).

@gushphp/contributor Do you have any comment on this idea?

**Comment by sstok** [comment]
@cordoval @pierredup @jubianchi ping

**Comment by cordoval** [comment]
i will start this

**Comment by pierredup** [comment]
@sstok I think having the adapters in the core repository makes a lot of sense, since it would make a lot of things easier. It would still be nice to be able to add external adapters as plugins, since I don't think gush should cater for absolutely every git hosting available, but can rather target the main ones and hav the community develop adapters for the ones they need

**Comment by sstok** [comment]
@pierredup OK, Thanks for the head-up :+1: 

For installing plug-ins we can actually have a look Puli, but I'm not sure how well the Composer support is.

**Comment by sstok** [comment]
One problem with merging them (back) into the core is loss of authorship and commits, we can't just merge/cherry-pick them as the paths are not correct and other files (README, LICENSE etc) also exist.

I can make a commit per adapter, mentioning the original authors (to give some credit) or take walk to kill myself git-amend rebasing every commit :skull:

**Comment by pierredup** [comment]
How about adding them as subtrees? That way you can keep the commit history and have the adapters as read-only repositories. I'm not sure what the maintenance is to keep a sub-tree in sync though

**Comment by cordoval** [comment]
For authors we can keep the composer.json or readme or some sort of info saying their names. They are close authors though so we can reach them at any moment. For the folder structure we can put them under ThirdParty folder.

**Comment by cordoval** [comment]
done
